File: agents/python3/agent_fwd.py
from typing import Union
from forward_model import ForwardModel
from game_state import GameState
import asyncio
import os
import random

fwd_model_uri = os.environ.get(
    "FWD_MODEL_CONNECTION_STRING") or "ws://127.0.0.1:6969/?role=admin"

uri = os.environ.get(
    "GAME_CONNECTION_STRING") or "ws://127.0.0.1:3000/?role=agent&agentId=agentId&name=defaultName"

actions = ["up", "down", "left", "right", "bomb", "detonate"]


class Agent():
    def __init__(self):
        self._client_fwd = ForwardModel(fwd_model_uri)
        self._client = GameState(uri)

        self._client.set_game_tick_callback(self._on_game_tick)
        self._client_fwd.set_next_state_callback(self._on_next_game_state)
        self.connect()

    def connect(self):
        loop = asyncio.get_event_loop()

        client_connection = loop.run_until_complete(self._client.connect())

        client_fwd_connection = loop.run_until_complete(
            self._client_fwd.connect())

        loop = asyncio.get_event_loop()
        loop.create_task(self._client._handle_messages(client_connection))
        loop.create_task(
            self._client_fwd._handle_messages(client_fwd_connection))
        loop.run_forever()

    def _get_bomb_to_detonate(self, game_state) -> Union[int, int] or None:
        agent_number = game_state.get("connection").get("agent_number")
        entities = self._client._state.get("entities")
        bombs = list(filter(lambda entity: entity.get(
            "owner") == agent_number and entity.get("type") == "b", entities))
        bomb = next(iter(bombs or []), None)
        if bomb != None:
            return [bomb.get("x"), bomb.get("y")]
        else:
            return None

    async def _on_game_tick(self, tick_number, game_state):
        await self._send_eval_next_state()
        random_action = self.generate_random_action()
        if random_action in ["up", "left", "right", "down"]:
            await self._client.send_move(random_action)
        elif random_action == "bomb":
            await self._client.send_bomb()
        elif random_action == "detonate":
            bomb_coordinates = self._get_bomb_to_detonate(game_state)
            if bomb_coordinates != None:
                x, y = bomb_coordinates
                await self._client.send_detonate(x, y)
        else:
            print(f"Unhandled action: {random_action}")

    async def _send_eval_next_state(self):
        actions = [
            {
                "action": {"move": "right", "type": "move"},
                "agent_number": 0,
            }, {
                "action": {"move": "left", "type": "move"},
                "agent_number": 1,
            }
        ]
        await self._client_fwd.send_next_state(1, self._client._state, actions)

    def generate_random_action(self):
        actions_length = len(actions)
        return actions[random.randint(0, actions_length - 1)]

    async def _on_next_game_state(self, state):
        # print(state)
        pass

    def generate_random_action(self):
        actions_length = len(actions)
        return actions[random.randint(0, actions_length - 1)]


def main():
    Agent()


if __name__ == "__main__":
    main()

File: agents/typescript/src/Agent_fwd.ts
import { AgentPacket, EntityType, GameStateClient, IGameState, PacketType, UnitMove } from "@coderone/bomberland-library";
import { ForwardAction, ForwardModel, NextStatePayload } from "./ForwardModel";

const gameConnectionString =
    process.env["GAME_CONNECTION_STRING"] || "ws://127.0.0.1:3000/?role=agent&agentId=agentIdA&name=RandomAgentFwd";
const fwdModelConnectionString = process.env["FWD_MODEL_CONNECTION_STRING"] || "ws://127.0.0.1:6969/?role=admin";

enum Action {
    Up = "up",
    Down = "down",
    Left = "left",
    Right = "right",
    Bomb = "bomb",
    Detonate = "detonate",
}

const actionMoveMap = new Map<Action, UnitMove>([
    [Action.Up, UnitMove.Up],
    [Action.Down, UnitMove.Down],
    [Action.Left, UnitMove.Left],
    [Action.Right, UnitMove.Right],
]);

const actionList = Object.values(Action);

class Agent {
    private readonly client = new GameStateClient(gameConnectionString);
    private readonly forwardModel = new ForwardModel(fwdModelConnectionString);
    private sequenceId = 0;

    public constructor() {
        this.client.SetGameTickCallback(this.onGameTick);
        this.forwardModel.SetNextStateCallback(this.onNextGameState);
    }

    private onGameTick = async (gameState: Omit<IGameState, "connection"> | undefined) => {
        if (gameState === undefined) {
            return;
        }

        // Forward-model demo: ask "what happens next tick if all my units move right?"
        // before committing to a real action. A real agent would evaluate several candidate
        // action sets and pick the best-scoring resulting state.
        this.evaluateNextState(gameState);

        const myAgentId = this.client.Connection?.agent_id ?? "";
        const units = gameState.agents[myAgentId]?.unit_ids ?? [];
        units.forEach((unitId) => {
            const action = this.generateAction();
            if (action) {
                const mappedMove = actionMoveMap.get(action);
                if (mappedMove !== undefined) {
                    this.client.SendMove(unitId, mappedMove);
                } else if (action === Action.Bomb) {
                    this.client.SendPlaceBomb(unitId);
                } else if (action === Action.Detonate) {
                    const bombCoordinates = this.getBombToDetonate(gameState);
                    if (bombCoordinates !== undefined) {
                        this.client.SendDetonateBomb(unitId, bombCoordinates);
                    }
                }
            }
        });
    };

    private evaluateNextState = (gameState: Omit<IGameState, "connection">) => {
        const myAgentId = this.client.Connection?.agent_id ?? "";
        const units = gameState.agents[myAgentId]?.unit_ids ?? [];
        const actions: Array<ForwardAction> = units.map((unitId) => {
            const action: AgentPacket = { type: PacketType.Move, move: UnitMove.Right, unit_id: unitId };
            return { agent_id: myAgentId, action };
        });
        this.forwardModel.SendEvaluateNextState(this.sequenceId++, gameState, actions);
    };

    private onNextGameState = (payload: NextStatePayload) => {
        // The forward model has simulated the hypothetical actions. `payload.next_state` is the
        // resulting state you would score in a search; here we just log a summary.
        console.log(
            `Forward model seq=${payload.sequence_id}: is_complete=${payload.is_complete}, ` +
                `resulting events=${payload.tick_result.events.length}, next tick=${payload.next_state.tick}`
        );
    };

    private generateAction = (): Action | undefined => {
        const allActions = actionList.length;
        const rand = Math.round(Math.random() * allActions);
        if (rand !== allActions) {
            return actionList[rand];
        }
    };

    private getBombToDetonate = (gameState: Omit<IGameState, "connection">): [number, number] | undefined => {
        const currentAgent = this.client.Connection?.agent_id;
        const bomb = gameState.entities.find((entity) => {
            const isBomb = entity.type === EntityType.Bomb;
            const isOwner = currentAgent !== undefined ? entity.agent_id === currentAgent : false;
            return isBomb === true && isOwner === true;
        });

        if (bomb?.x !== undefined && bomb.y !== undefined) {
            return [bomb.x, bomb.y];
        }
    };
}

new Agent();
